File: src/physics/physics.cpp
/* Lips of Suna
 * Copyright© 2007-2009 Lips of Suna development team.
 *
 * Lips of Suna is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Lips of Suna is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Lips of Suna. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * \addtogroup liphy Physics
 * @{
 * \addtogroup liphyPhysics Physics
 * @{
 */

#include "physics.h"
#include "physics-private.h"

#define MAXPROXIES 1024

static void
private_internal_tick (btDynamicsWorld* dynamics,
                       btScalar         step);

/*****************************************************************************/

/**
 * \brief Creates a new physics simulation.
 *
 * \return New physics simulation or NULL.
 */
liphyPhysics*
liphy_physics_new ()
{
	liphyPhysics* self;
	btVector3 min (-65535, -65535, -65535);
	btVector3 max ( 65535,  65535,  65535);

	self = (liphyPhysics*) calloc (1, sizeof (liphyPhysics));
	if (self == NULL)
		return NULL;
	try
	{
		self->broadphase = new bt32BitAxisSweep3 (min, max, MAXPROXIES);
		self->broadphase->getOverlappingPairCache ()->setInternalGhostPairCallback (new btGhostPairCallback ());
		self->configuration = new btDefaultCollisionConfiguration ();
		self->dispatcher = new btCollisionDispatcher (self->configuration);
		self->solver = new btSequentialImpulseConstraintSolver ();
		self->dynamics = new btDiscreteDynamicsWorld (self->dispatcher, self->broadphase, self->solver, self->configuration);
		self->dynamics->setGravity (btVector3 (0.0f, -10.0f, 0.0f));
		self->dynamics->setInternalTickCallback (private_internal_tick, self);
	}
	catch (...)
	{
		liphy_physics_free (self);
		return NULL;
	}

	return self;
}

/**
 * \brief Frees the physics simulation.
 *
 * \param self Physics simulation.
 */
void
liphy_physics_free (liphyPhysics* self)
{
	delete self->dynamics;
	delete self->solver;
	delete self->configuration;
	delete self->dispatcher;
	delete self->broadphase;
	lialg_list_free (self->controllers);
	free (self);
}

/**
 * \brief Casts a ray to the scene.
 *
 * If the ray hits no obstacle, result is set to the ray end point, normal is
 * set to ray direction, and zero is returned. Otherwise, nonzero is returned
 * and result and normal hold information on the collision point.
 *
 * \param self Physics simulation.
 * \param start Ray start point.
 * \param end Ray end point.
 * \param result Return location for collision point.
 * \param normal Return location for collision surface normal.
 * \return Nonzero if no collision occurred.
 */
int
liphy_physics_cast_ray (const liphyPhysics* self,
                        const limatVector*     start,
                        const limatVector*     end,
                        limatVector*           result,
                        limatVector*           normal)
{
	btVector3 src (start->x, start->y, start->z);
	btVector3 dst (end->x, end->y, end->z);
//	btBroadphaseProxy* proxy;
	btCollisionWorld* collision;
	btCollisionWorld::ClosestRayResultCallback test (src, dst);

	/* Cast the ray. */
//	proxy = self->body->getBroadphaseHandle ();
	collision = self->dynamics->getCollisionWorld ();
	test.m_closestHitFraction = 1.0f;
//	test.m_collisionFilterGroup = proxy->m_collisionFilterGroup;
//	test.m_collisionFilterMask = proxy->m_collisionFilterMask;
	collision->rayTest (src, dst, test);

	/* Inspect results. */
	if (test.m_closestHitFraction < 1.0f)
	{
		*result = limat_vector_init (test.m_hitPointWorld[0], test.m_hitPointWorld[1], test.m_hitPointWorld[2]);
		*normal = limat_vector_init (test.m_hitNormalWorld[0], test.m_hitNormalWorld[1], test.m_hitNormalWorld[2]);
		return 0;
	}
	else
	{
		*result = *end;
		*normal = limat_vector_normalize (limat_vector_subtract (*end, *start));
		return 1;
	}
}

/**
 * \brief Updates the physics simulation.
 *
 * \param self Physics simulation.
 * \param secs Tick length in seconds.
 */
void
liphy_physics_update (liphyPhysics* self,
                      float         secs)
{
	self->dynamics->stepSimulation (secs, 20);
}

/**
 * \brief Sets the object transformation callback.
 *
 * The callback will be called every time an object is moved.
 *
 * \param self Physics simulation.
 * \param func Transformation callback function.
 */
void
liphy_physics_set_transform_callback (liphyPhysics*      self,
                                      liphyTransformFunc func)
{
	self->callbacks.transform = func;
}

/*****************************************************************************/

static void
private_internal_tick (btDynamicsWorld* dynamics,
                       btScalar         step)
{
#if 0
	lialgList* ptr;
	liphyObject* object;
	liphyPhysics* self = (liphyPhysics*) dynamics->getWorldUserInfo ();

	for (ptr = self->controllers ; ptr != NULL ; ptr = ptr->next)
	{
		object = (liphyObject*) ptr->data;
		assert (object->control.call0 != NULL || object->control.call1 != NULL);
		if (object->control.call0 != NULL)
			object->control.call0 (object, step);
		if (object->control.call1 != NULL)
			object->control.call1 (object, step);
	}
#endif
}

/** @} */
/** @} */

File: data/system/physics.lua
--- Provides access to the physics simulation.
--
-- Lips of Suna is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Lesser General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License, or (at your option) any later version.
--
-- @module system.physics
-- @alias Physics

local Class = require("system/class")
local Vector = require("system/math/vector")

if not Los.program_load_extension("physics") then
	error("loading extension `physics' failed")
end

------------------------------------------------------------------------------

--- Provides access to the physics simulation.
-- @type Physics
local Physics = Class("Physics")

--- Performs a ray cast test.
-- @param self Physics class.
-- @param src Start point vector in world space.
-- @param dst End point vector in world space.
-- @param mask Collision mask.
-- @param ignore Table of physics objects to ignore. Nil to not ignore any.
-- @return Table with point, normal, object and tile. Nil if no collision occurred.
Physics.cast_ray = function(self, src, dst, mask, ignore)
	local ignore_
	if ignore then
		ignore_ = {}
		for k,v in pairs(ignore) do ignore_[k] = v.handle end
	end
	local r = Los.physics_cast_ray(src.handle, dst.handle, mask, ignore_)
	if not r then return end
	r.point = Vector:new_from_handle(r.point)
	r.normal = Vector:new_from_handle(r.normal)
	if r.tile then r.tile = Vector:new_from_handle(r.tile) end
	return r
end

--- Performs a sphere cast test.
-- @param self Physics class.
-- @param src Start point vector in world space.
-- @param dst End point vector in world space.
-- @param radius Radius in world units.
-- @param mask Collision mask.
-- @param ignore Table of physics objects to ignore. Nil to not ignore any.
-- @return Table with point, normal, object and tile. Nil if no collision occurred.
Physics.cast_sphere = function(self, src, dst, radius, mask, ignore)
	local ignore_
	if ignore then
		ignore_ = {}
		for k,v in pairs(ignore) do ignore_[k] = v.handle end
	end
	local r = Los.physics_cast_sphere(src.handle, dst.handle, radius, mask, ignore_)
	if not r then return end
	r.point = Vector:new_from_handle(r.point)
	r.normal = Vector:new_from_handle(r.normal)
	if r.tile then r.tile = Vector:new_from_handle(r.tile) end
	return r
end

--- Gets the physics simulation enable status.
-- @param self Physics class.
-- @return Boolean.
Physics.get_enable_simulation = function(self)
	return Los.physics_get_enable_simulation()
end

--- Sets the physics simulation enable status.
-- @param self Physics class.
-- @param v Boolean.
Physics.set_enable_simulation = function(self, v)
	Los.physics_set_enable_simulation(v)
end

return Physics
